cdap-data-fabric/src/main/java/co/cask/cdap/data2/metadata/dataset/MdsHistoryKey.java

```diff
@@ -51,33 +46,37 @@ static long extractTime(byte[] rowKey, String type) {
     // Skip rowType
     keySplitter.skipBytes();
 
-    // Skip targetId
-    if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ProgramId.class))) {
-      keySplitter.skipString();
-      keySplitter.skipString();
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ApplicationId.class))) {
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(DatasetId.class))) {
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(StreamId.class))) {
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(StreamViewId.class))) {
-      // skip namespace, stream, view
-      keySplitter.skipString();
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else if (type.equals(EntityIdKeyHelper.TYPE_MAP.get(ArtifactId.class))) {
-      // skip namespace, name, version
-      keySplitter.skipString();
-      keySplitter.skipString();
-      keySplitter.skipString();
-    } else {
-      throw new IllegalArgumentException("Illegal Type " + type + " of metadata source.");
+    switch (type) {
+      case MetadataEntity.PROGRAM:
```

Comment:
you can further simplify this by merging all cases that skip two strings into one:

```
case APPLICATION:
case DATASET:
case STREAM:
    skipString();
    skipString();
    break;
case VIEW:
case ARTIFACT:
    skipString();
    skipString();
    skipString();
    break;
case PROGRAM:
    skipString();
    skipString();
    skipString();
    skipString();
    break;
default: ....
```